README.md
# discovery-swarm-webrtc

[![Build Status](https://travis-ci.com/geut/discovery-swarm-webrtc.svg?branch=master)](https://travis-ci.com/geut/discovery-swarm-webrtc)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

> discovery-swarm for webrtc

<p align="center">
  <img src="https://user-images.githubusercontent.com/819446/64871056-d6a2d480-d61a-11e9-9d93-b79a5f0e822a.gif" alt="force-graph">
</p>

This module provides a "similar" API to discovery-swarm but for WebRTC connections.

It has a few differences to discovery-swarm:

- It needs a signaling server. We give you [one](#server).
- It uses [mmst](https://github.com/RangerMauve/mostly-minimal-spanning-tree) to minimize the connections. Check the example.
- `join` and `leave` only accepts Buffers.
- `leave` and `close` accepts a callback argument or returns a Promise.

## Install

```
$ npm install @geut/discovery-swarm-webrtc
```

## Usage

### <a name="server"></a>Server

You can run your own signal server by running:

```
$ discovery-swarm-webrtc --port=4000
```

#### Public Servers

- wss://geut-webrtc-signal-v3.glitch.me
- wss://geut-webrtc-signal-v3.herokuapp.com

#### Deploy to Heroku

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/geut/discovery-swarm-webrtc/tree/master)

### Client

```javascript
const crypto = require('crypto')
const swarm = require('@geut/discovery-swarm-webrtc')

const sw = swarm({
  bootstrap: ['ws://localhost:4000']
})

const topic = crypto.createHash('sha256')
  .update('my-discovery-swarm-topic')
  .digest()

sw.join(topic)

sw.on('connection', peer => {
  // connected
})
```

## API

#### `const sw = swarm(opts)`

Creates a new Swarm.

`opts` include:

```javascript
{
  id: crypto.randomBytes(32), // peer-id for user
  bootstrap: [string], // urls to your websocket endpoints
  stream: (info) => stream, // stream to replicate across peers
  simplePeer: {}, // options for the simplePeer instances,
  maxPeers: 5, // max connections by peer
  timeout: 15 * 1000, // defines the time to wait to establish a connection
}
```

#### `sw.join(Buffer)`

Join a specific channel. We use behind it `simple-signal` + `simple-peer`.

#### `const promise = sw.leave(Buffer)`

Leave from specific channel. Destroy all the connections and leave the channel.

#### `const promise = sw.close([callback])`

Close the entire swarm. Destroy all the connections and disconnect from the signal.

#### `const arrayOfPeers = sw.getPeers([channel])`

Returns the list of connected peers for a specific channel.

Channel is `optional`, if you don't pass it you get the entire list of peers.

#### `sw.connect(channel: Buffer, peerId: Buffer) -> Promise<SimplePeer>`

Connect directly to a specific peer.

### Events

#### `sw.on('handshaking', function(connection, info) { ... })`

Emitted when you've connected to a peer and are now initializing the connection's session. Info is an object that contains information about the connection.

`info` include:

``` js
{
  id: Buffer // the remote peer's peer-id.
  channel: Buffer // the channel this connection was initiated on.
  initiator: Boolean // whether we initiated the connection or someone else did
}
```

#### `sw.on('connection', function(connection, info) { ... })`

Emitted when you have fully connected to another peer. Info is an object that contains info about the connection.

#### `sw.on('connection-closed', function(connection, info) { ... })`

Emitted when you've disconnected from a peer. Info is an object that contains info about the connection.

#### `sw.on('leave', function(channel) { ... })`

Emitted when you left a channel.

#### `sw.on('close', function() { ... })`

Emitted when the swarm was closed.

#### `sw.on('candidates-updated', function(channel, candidates) { ... })`

Emitted when the candidates peer for a specific channel was updated. `candidates` is an array of Buffer id.

## <a name="issues"></a> Issues

:bug: If you found an issue we encourage you to report it on [github](https://github.com/geut/discovery-swarm-webrtc/issues). Please specify your OS and the actions to reproduce it.

## <a name="contribute"></a> Contributing

:busts_in_silhouette: Ideas and contributions to the project are welcome. You must follow this [guideline](https://github.com/geut/discovery-swarm-webrtc/blob/master/CONTRIBUTING.md).

## Sponsored By
[
<img src="https://dxos.org/images/logo/dxos-logo-type-dark.png" alt="DXOS" width="350px" />
](https://dxos.org/)

## License

MIT © A [**GEUT**](http://geutstudio.com/) project


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 4.3.1 - 2021-07-12 [YANKED]

## 4.3.0 - 2021-04-22 [YANKED]

## 4.2.2 - 2021-02-25 [YANKED]

## 4.2.1 - 2021-02-10 [YANKED]

## 4.2.0 - 2021-02-04 [YANKED]

## 4.1.3 - 2021-01-29 [YANKED]

## 4.1.2 - 2021-01-29 [YANKED]

## 4.1.1 - 2021-01-27 [YANKED]

## 4.1.0 - 2021-01-21 [YANKED]

## 4.0.2 - 2020-10-08 [YANKED]

## 4.0.1 - 2020-09-03 [YANKED]

## 4.0.0 - 2020-08-21 [YANKED]

## [3.0.0] - 2020-08-21
### Added
- Replace socket.io/simple-signal in favor of socket-signal

## 2.2.7 - 2020-01-22 [YANKED]

## 2.2.6 - 2020-01-22 [YANKED]

## 2.2.5 - 2020-01-08 [YANKED]

## 2.2.4 - 2019-10-16 [YANKED]

## 2.2.3 - 2019-09-22 [YANKED]

## 2.2.2 - 2019-09-19 [YANKED]

## 2.2.1 - 2019-09-18 [YANKED]

## [2.2.0] - 2019-09-17
### Added
- Callback support for leave and close.

## 2.1.1 - 2019-09-16 [YANKED]

## [2.1.0] - 2019-09-16
### Changed
- Refactor signal server to work with buffers.

### Fixed
- Peer leak on error connection.

## [2.0.0] - 2019-09-13
### Added
- mmst algorhitm to reduce and optimize WebRTC connections.
- timeout for connections and requests.
- visual example using ngraph and force-graph.
- connectivity test using ngraph.

### Changed
- move signal websocket logic outside to provide a clean signal API.

## 1.0.0 - 2019-09-11 [YANKED]
[Unreleased]: https://github.com/geut/discovery-swarm-webrtc/compare/v4.3.1...HEAD
[3.0.0]: https://github.com/geut/discovery-swarm-webrtc/compare/v2.2.7...v3.0.0
[2.2.0]: https://github.com/geut/discovery-swarm-webrtc/compare/v2.1.1...v2.2.0
[2.1.0]: https://github.com/geut/discovery-swarm-webrtc/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/geut/discovery-swarm-webrtc/compare/v1.0.0...v2.0.0


index.js
const { EventEmitter } = require('events')
const crypto = require('crypto')
const assert = require('nanocustomassert')

const pump = require('pump')

const log = require('debug')('discovery-swarm-webrtc')
const MMSTSignalClient = require('./lib/mmst-signal-client')
const { toHex, callbackPromise, resolveCallback } = require('./lib/utils')
const errors = require('./lib/errors')

const { ERR_CONNECTION_DUPLICATED } = errors

const assertChannel = channel => assert(Buffer.isBuffer(channel) && channel.length === 32, 'Channel must be a buffer of 32 bytes')

class DiscoverySwarmWebrtc extends EventEmitter {
  constructor (opts = {}) {
    super()
    log('opts', opts)

    const { id = crypto.randomBytes(32), bootstrap, stream, simplePeer, maxPeers = 5, timeout = 15 * 1000, signal, mmst = {} } = opts

    assert(Array.isArray(bootstrap) && bootstrap.length > 0, 'The `bootstrap` options is required.')
    assert(Buffer.isBuffer(id) && id.length === 32, 'The `id` option needs to be a Buffer of 32 bytes.')

    this.id = id

    const queueTimeout = timeout * 2

    const signalOptions = {
      id: this.id,
      bootstrap,
      createConnection: peer => this._createConnection(peer),
      requestTimeout: timeout,
      simplePeer,
      reconnectingWebsocket: {
        connectionTimeout: timeout
      }
    }

    this.signal = signal ? signal(signalOptions, this) : new MMSTSignalClient({
      ...signalOptions,
      mmstOpts: {
        maxPeers,
        queueTimeout, // queue mmst
        lookupTimeout: timeout,
        ...mmst
      }
    })

    this._maxPeers = maxPeers
    this._stream = stream
    this._destroyed = false

    this._initialize(opts)
  }

  get connected () {
    return this.signal.peersConnected
  }

  get connecting () {
    return this.signal.peersConnecting
  }

  listen () {
    // Empty method to respect the API of discovery-swarm
  }

  getPeers (channel) {
    if (channel) {
      return this.signal
        .getPeersByTopic(channel)
        .filter(p => p.connected)
    }
    return this.signal.peersConnected
  }

  join (channel) {
    this.signal.join(channel)
  }

  leave (channel, cb = callbackPromise()) {
    assertChannel(channel)
    resolveCallback(this.signal.leave(channel), cb)
    return cb.promise
  }

  close (cb = callbackPromise()) {
    resolveCallback(this._close(), cb)
    return cb.promise
  }

  async connect (channel, peerId) {
    assert(channel && Buffer.isBuffer(channel), 'channel must be a buffer')
    assert(peerId && Buffer.isBuffer(peerId), 'peerId must be a buffer')

    const peer = this.signal.connect(channel, peerId)
    peer.subscribeMediaStream = true
    await peer.ready()
    return peer.stream
  }

  async _close () {
    if (this._destroyed) return
    this._destroyed = true
    await this.signal.close()
    this.emit('close')
  }

  _initialize () {
    const signal = this.signal

    // It would log the errors and prevent of throw it.
    this.on('error', (...args) => log('error', ...args))

    signal.on('peer-error', err => this.emit('error', err))
    signal.on('error', err => this.emit('error', err))
    signal.open().catch(err => process.nextTick(() => this.emit('error', err)))
    signal.on('candidates-updated', (...args) => this.emit('candidates-updated', ...args))
  }

  _createConnection (peer) {
    peer.subscribeMediaStream = true

    peer.channel = peer.topic

    peer.getInfo = () => ({
      id: peer.id,
      channel: peer.topic,
      initiator: peer.initiator
    })

    peer.printInfo = () => ({
      id: toHex(peer.id),
      channel: toHex(peer.topic),
      initiator: peer.initiator
    })

    log('createConnection', { info: peer.printInfo() })

    try {
      const duplicate = this._checkForDuplicate(peer)
      if (duplicate) {
        if (duplicate === peer) {
          // current incoming connection
          throw new ERR_CONNECTION_DUPLICATED(toHex(this.id), toHex(peer.id))
        } else {
          // old connection
          duplicate.destroy()
        }
      }

      this._bindSocketEvents(peer)

      return peer
    } catch (err) {
      this.emit('connect-failed', err, peer.getInfo())
      throw err
    }
  }

  _bindSocketEvents (peer) {
    const info = peer.getInfo()

    peer.on('error', err => {
      log('error', err)
      this.emit('connection-error', err, info)
    })

    peer.on('connect', () => {
      log('connect', { peer })
      if (peer.stream.destroyed) {
        return
      }

      if (!this._stream) {
        this._handleConnection(peer.stream, info)
        return
      }

      const conn = this._stream(info)
      if (!conn) {
        this._handleConnection(peer.stream, info)
        return
      }

      this.emit('handshaking', conn, info)
      conn.on('handshake', this._handshake.bind(this, conn, info))
      pump(peer.stream, conn, peer.stream)
    })

    peer.on('close', () => {
      log('close', { peer })
      this.emit('connection-closed', peer.stream, info)
    })
  }

  _handshake (conn, info) {
    this._handleConnection(conn, info)
  }

  _handleConnection (conn, info) {
    this.emit('connection', conn, info)
  }

  _checkForDuplicate (peer) {
    const oldPeer = this.signal.getPeersByTopic(peer.channel).find(p => p.id.equals(peer.id) && !p.sessionId.equals(peer.sessionId))
    if (!oldPeer) {
      return
    }

    const connections = [peer, oldPeer]

    /**
     * The first case is to have duplicate connections from the same origin (remote or local).
     * In this case we do a sort by connectionId and destroy the first one.
     */
    if ((peer.initiator && oldPeer.initiator) || (!peer.initiator && !oldPeer.initiator)) {
      return connections.sort((a, b) => Buffer.compare(a.sessionId, b.sessionId))[0]
    }

    /**
     * The second case is to have duplicate connections where each connection is started from different origins.
     * In this case we do a sort by peer id and destroy the first one.
     */
    const toDestroy = [this.id, peer.id].sort(Buffer.compare)[0]
    return connections.find(p => p.id.equals(toDestroy))
  }
}

module.exports = (...args) => new DiscoverySwarmWebrtc(...args)
module.exports.errors = errors
